src/auth/src/credentials/mds_credential.rs

```diff
@@ -13,37 +13,164 @@
+    scopes: Option<Vec<String>>,
+    universe_domain: Option<String>,
+}
+
+#[allow(dead_code)]
+impl MDSCredentialBuilder {
+    pub fn endpoint<S: Into<String>>(mut self, endpoint: S) -> Self {
+        self.endpoint = Some(endpoint.into());
+        self
+    }
+
+    pub fn quota_project_id<S: Into<String>>(mut self, quota_project_id: S) -> Self {
+        self.quota_project_id = Some(quota_project_id.into());
+        self
+    }
+
+    pub fn universe_domain<S: Into<String>>(mut self, universe_domain: S) -> Self {
+        self.universe_domain = Some(universe_domain.into());
+        self
+    }
+
+    pub fn scopes<S: Into<String>>(mut self, scopes: Vec<S>) -> Self {
+        self.scopes = Some(scopes.into_iter().map(|s| s.into()).collect());
+        self
+    }
+
+    pub fn build(&self) -> Credential {
+        let endpoint = self.endpoint.clone().unwrap_or(METADATA_ROOT.to_string());
+
+        let token_provider = MDSAccessTokenProvider::builder()
+            .endpoint(endpoint.clone())
+            .maybe_scopes(self.scopes.clone())
+            .build();
+
+        let ud_cache = UniverseDomainCache {
+            universe_domain: self.universe_domain.clone(),
+            is_cached: self.universe_domain.is_some(),
+        };
+        let (ud_tx, ud_rx) = watch::channel(ud_cache);
+        let notify = Arc::new(Notify::new());
+
+        let mdsc = MDSCredential {
+            quota_project_id: self.quota_project_id.clone(),
+            token_provider,
+            ud_rx,
+            wakeup_tx: notify.clone(),
+        };
+
+        if self.universe_domain.is_none() {
+            tokio::spawn(async move {
+                MDSCredential::<MDSAccessTokenProvider>::universe_domain_background_task(
+                    endpoint,
+                    ud_tx,
+                    notify.clone(),
+                )
+                .await;
+            });
+        }
+        Credential {
+            inner: Arc::new(mdsc),
+        }
+    }
+}
+
+impl<T> MDSCredential<T>
+where
+    T: TokenProvider,
+{
+    async fn get_universe_domain_from_mds(endpoint: String) -> Option<String> {
+        let client = Client::new();
+        let request = client
+            .get(format!("{}/universe/universe-domain", endpoint))
+            .header(
+                METADATA_FLAVOR,
+                HeaderValue::from_static(METADATA_FLAVOR_VALUE),
+            );
+
+        let response = request.send().await.ok()?;
+
+        if !response.status().is_success() {
+            let status = response.status();
+            if status == StatusCode::NOT_FOUND {
+                return Some(DEFAULT_UNIVERSE_DOMAIN.to_string());
+            }
+            return None;
+        }
+        let universe_domain = response
+            .json::<UniverseDomainResponse>()
+            .await
+            .ok()?
+            .universe_domain;
+
+        Some(universe_domain)
+    }
+
+    async fn universe_domain_background_task(
```

Comment:
Why is this a member of `MDSCredential<T>`? It does not use `Self`, and it seems to be called from outside the struct only?